From 16ff55d501af Mon Sep 17 00:00:00 2001
From: Suren Baghdasaryan <surenb@google.com>
Date: Mon Aug 24 18:32:48 2026 +0000
Subject: [PATCH] ANDROID: avoid ABI breakage in struct psi_trigger

ANDROID: avoid ABI breakage in struct psi_trigger

struct psi_trigger is part of the GKI ABI, so the field added by 006
would break it. Move it into an external psi_trigger_ext container and
allocate that instead.

Apply after 006.

Source: AOSP kernel/common android14-6.1 16ff55d501af
---
diff --git a/include/linux/psi_types.h b/include/linux/psi_types.h
index 1357fd1..bb8b7c1 100644
--- a/include/linux/psi_types.h
+++ b/include/linux/psi_types.h
@@ -134,9 +134,6 @@ struct psi_trigger {
 	/* Wait queue for polling */
 	wait_queue_head_t event_wait;
 
-	/* Kernfs file for cgroup triggers */
-	struct kernfs_open_file *of;
-
 	/* Pending event flag */
 	int event;
 
@@ -153,6 +150,14 @@ struct psi_trigger {
 	bool pending_event;
 };
 
+struct psi_trigger_ext {
+	struct psi_trigger trigger;
+
+	/* Kernfs file for cgroup triggers */
+	struct kernfs_open_file *of;
+
+};
+
 struct psi_group {
 	struct psi_group *parent;
 	bool enabled;
diff --git a/kernel/sched/psi.c b/kernel/sched/psi.c
index 598eefed..dd7f31b 100644
--- a/kernel/sched/psi.c
+++ b/kernel/sched/psi.c
@@ -555,8 +555,11 @@ static u64 update_triggers(struct psi_group *group, u64 now)
 
 		/* Generate an event */
 		if (cmpxchg(&t->event, 0, 1) == 0) {
-			if (t->of)
-				kernfs_notify(t->of->kn);
+			struct psi_trigger_ext *t_ext;
+
+			t_ext = container_of(t, struct psi_trigger_ext, trigger);
+			if (t_ext->of)
+				kernfs_notify(t_ext->of->kn);
 			else
 				wake_up_interruptible(&t->event_wait);
 		}
@@ -1252,6 +1255,7 @@ struct psi_trigger *psi_trigger_create(struct psi_group *group, char *buf,
 				       enum psi_res res, struct file *file,
 				       struct kernfs_open_file *of)
 {
+	struct psi_trigger_ext *t_ext;
 	struct psi_trigger *t;
 	enum psi_states state;
 	u32 threshold_us;
@@ -1282,9 +1286,10 @@ struct psi_trigger *psi_trigger_create(struct psi_group *group, char *buf,
 	if (threshold_us == 0 || threshold_us > window_us)
 		return ERR_PTR(-EINVAL);
 
-	t = kmalloc(sizeof(*t), GFP_KERNEL);
-	if (!t)
+	t_ext = kmalloc(sizeof(*t_ext), GFP_KERNEL);
+	if (!t_ext)
 		return ERR_PTR(-ENOMEM);
+	t = &t_ext->trigger;
 
 	t->group = group;
 	t->state = state;
@@ -1295,7 +1300,7 @@ struct psi_trigger *psi_trigger_create(struct psi_group *group, char *buf,
 
 	t->event = 0;
 	t->last_event_time = 0;
-	t->of = of;
+	t_ext->of = of;
 	if (!of)
 		init_waitqueue_head(&t->event_wait);
 	t->pending_event = false;
@@ -1307,7 +1312,7 @@ struct psi_trigger *psi_trigger_create(struct psi_group *group, char *buf,
 
 		task = kthread_create(psi_poll_worker, group, "psimon");
 		if (IS_ERR(task)) {
-			kfree(t);
+			kfree(t_ext);
 			mutex_unlock(&group->trigger_lock);
 			return ERR_CAST(task);
 		}
@@ -1329,6 +1334,7 @@ struct psi_trigger *psi_trigger_create(struct psi_group *group, char *buf,
 
 void psi_trigger_destroy(struct psi_trigger *t)
 {
+	struct psi_trigger_ext *t_ext;
 	struct psi_group *group;
 	struct task_struct *task_to_destroy = NULL;
 
@@ -1345,8 +1351,9 @@ void psi_trigger_destroy(struct psi_trigger *t)
 	 * being accessed later. Can happen if cgroup is deleted from under a
 	 * polling process.
 	 */
-	if (t->of)
-		kernfs_notify(t->of->kn);
+	t_ext = container_of(t, struct psi_trigger_ext, trigger);
+	if (t_ext->of)
+		kernfs_notify(t_ext->of->kn);
 	else
 		wake_up_interruptible(&t->event_wait);
 
@@ -1396,13 +1403,14 @@ void psi_trigger_destroy(struct psi_trigger *t)
 		kthread_stop(task_to_destroy);
 		atomic_set(&group->poll_scheduled, 0);
 	}
-	kfree(t);
+	kfree(t_ext);
 }
 
 __poll_t psi_trigger_poll(void **trigger_ptr,
 				struct file *file, poll_table *wait)
 {
 	__poll_t ret = DEFAULT_POLLMASK;
+	struct psi_trigger_ext *t_ext;
 	struct psi_trigger *t;
 
 	if (static_branch_likely(&psi_disabled))
@@ -1412,8 +1420,9 @@ __poll_t psi_trigger_poll(void **trigger_ptr,
 	if (!t)
 		return DEFAULT_POLLMASK | EPOLLERR | EPOLLPRI;
 
-	if (t->of)
-		kernfs_generic_poll(t->of, wait);
+	t_ext = container_of(t, struct psi_trigger_ext, trigger);
+	if (t_ext->of)
+		kernfs_generic_poll(t_ext->of, wait);
 	else
 		poll_wait(file, &t->event_wait, wait);
 
